doc/source/dev/api-dev/array_api.rst

```diff
@@ -0,0 +1,253 @@
+    ...
+    TypeError: 'numpy.ma.MaskedArray' are not supported
+
+    >>> vq(features.astype(np.object_), code_book)
+    ...
+    TypeError: object arrays are not supported
+
+
+Currently supported functionality
+`````````````````````````````````
+
+The following modules provide Array API standard support when the environment
+variable is set:
+
+- ``scipy.cluster.hierarchy``
+- ``scipy.cluster.vq``
+
+
+Implementation notes
+--------------------
+
+A key part of the support for the array API standard and specific compatibility
+functions for Numpy, CuPy and PyTorch is provided through
+`array-api-compat <https://github.com/data-apis/array-api-compat>`_.
+This package is included in the SciPy code base via a git submodule (under
+``scipy/_lib``), so no new dependencies are introduced.
+
+``array-api_compat`` provides generic utility functions and adds aliases such
+as ``xp.concat`` (which, for numpy, maps to ``np.concatenate``). This allows
+using a uniform API across NumPy, PyTorch and CuPy (as of right now; support
+for other libraries like JAX is expected to be added in the future).
+
+When the environment variable isn't set and hence Array API support in SciPy is
+disabled, we still use the "augmented" version of the NumPy namespace, which is
+``array_api_compat.numpy`` (this should not change behavior of SciPy functions).
+When support is enabled, depending on the type of arrays, ``xp`` will return the
+standard-compatible namespace matching the input array type to a function (e.g.,
+if the input to ``cluster.vq.kmeans`` is a PyTorch array, then ``xp`` is
+``array_api_compat.torch``).
+
+
+Adding Array API support to a SciPy function
+--------------------------------------------
+
+As much as possible, new code added to SciPy should try to follow as closely as
+possible the Array API standard (these functions typically are best-practice
+idioms for NumPy usage as well). By following the standard, effectively adding
+support for Array API is typically straightforward, and we ideally don't need
+to maintain any customization.
+
+Two helper functions are available:
+
+* ``array_namespace``: detect the namespace based on input arrays and do some
+  input validation (like refusing to work with masked arrays, please see the
+  `RFC`_.)
+* ``as_xparray``: a drop-in replacement for ``np.asarray`` with additional
+  features like ``copy, check_finite``. As stated above, try to limit the use
+  of non standard features. In the end we would want to upstream our needs to
+  the compatibility library.
+
+To add support to a Python-level SciPy function, what you have to change is:
+
+1. Input array validation,
+2. Using ``xp`` rather ``np`` functions,
+3. When calling into compiled code, convert the array to a NumPy array before
+   and convert it back to the input array type after.
+
+Input array validation uses the following pattern::
+
+   xp = array_namespace(arr)  # where `arr` is the first input array
```

Comment:
What if a function has 2 array arguments and a user gives a list to the first and a pytorch tensor to the second?

Reply:
Only Array API compatible arrays are supported, meaning that if you pass a list (does not matter if you pass one or n arrays), it will get rejected.

The case of mixed type is also not supported and would raise. e.g. passing a CuPy array and a PyTorch tensor.

Reply:
We can spell this out more explicitly in the guide perhaps, since it's been asked about multiple times. For folks familiar with PyTorch, CuPy etc. this makes perfect sense, but NumPy has long instilled an "anything goes" expectation into its users - and we're trying to get away from that, or at least not make it worse.

Reply:
I am adding this to the docs:

*Note that if one input is a non-numpy array type, all array-like inputs have to
be of that type; trying to mix non-numpy arrays with lists, Python scalars or
other arbitrary Python objects will raise an exception. For NumPy arrays, those
types will continue to be accepted for backwards compatibility reasons.*